### PR TITLE
Ogr total feature count

### PR DESCRIPTION
**Pull request by tschmetzer**
Heavily speed up QgsOgrLayer::GetApproxFeatureCount() using readily available meta data instead of expensive individual one-by-one feature counting. Relevant for layers with more than a couple of hundred thousand features but really stands out for large layers with hundreds of millions of features.

### CODE REVIEW

File: src/core/providers/ogr/qgsogrproviderutils.cpp [hunk 1]
```diff
@@ -2909,6 +2927,46 @@ GIntBig QgsOgrLayer::GetApproxFeatureCount()
...
+
+GIntBig QgsOgrLayer::getTotalFeatureCountfromMetaData() const
+{
+  // Don't quote column name (see https://trac.osgeo.org/gdal/ticket/5799#comment:9)
+  QByteArray layerName = OGR_L_GetName( hLayer );
+
+  QString driverName = GDALGetDriverShortName( GDALGetDatasetDriver( ds->hDS ) );
+  if ( driverName == QLatin1String( "GPKG" ) )
+  {
+    // QByteArray sql = "SELECT COUNT (*) FROM " + QgsOgrProviderUtils::quotedIdentifier( layerName, driverName );
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrproviderutils.cpp:2942]: please remove that commented line of code and the other ones below

Comment (tschmetzer) [reply src/core/providers/ogr/qgsogrproviderutils.cpp]: done

**Comment by tschmetzer** [comment]
Can we reconsider the PR? What is blocking it and what can I do to unblock?

File: src/core/providers/ogr/qgsogrproviderutils.cpp [hunk 2]
```diff
@@ -2826,80 +2826,98 @@ GIntBig QgsOgrLayer::GetApproxFeatureCount()
...
-    OGRLayerH hSqlLayer = GDALDatasetExecuteSQL(
-                            ds->hDS, "SELECT 1 FROM gpkg_ogr_contents LIMIT 0", nullptr, nullptr );
-    CPLPopErrorHandler();
-    if ( hSqlLayer )
-    {
-      GDALDatasetReleaseResultSet( ds->hDS, hSqlLayer );
-      return OGR_L_GetFeatureCount( hLayer, TRUE );
-    }
+    // use feature count from meta data
+    GIntBig totalFeatureCount = getTotalFeatureCountfromMetaData();
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrproviderutils.cpp:2830]:
```suggestion
    GIntBig totalFeatureCount = GetTotalFeatureCountFromMetaData();
```

File: src/core/providers/ogr/qgsogrproviderutils.h [hunk 3]
```diff
@@ -446,6 +446,9 @@ class QgsOgrLayer
     //! Return an approximate feature count
     GIntBig GetApproxFeatureCount();
 
+    //! Return an total feature count based on meta data from package container
+    GIntBig getTotalFeatureCountfromMetaData() const;
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrproviderutils.h:450]:
```suggestion
    GIntBig GetTotalFeatureCountFromMetaData() const;
```

File: src/core/providers/ogr/qgsogrproviderutils.cpp [hunk 4]
```diff
@@ -2909,6 +2927,41 @@ GIntBig QgsOgrLayer::GetApproxFeatureCount()
   return OGR_L_GetFeatureCount( hLayer, TRUE );
 }
 
+/**
+ * Returns the total (unfiltered) feature count according to OGRGeoPackageTableLayer::GetFeatureCount used by GDAL's ogrinfo.
+ */
+
+GIntBig QgsOgrLayer::getTotalFeatureCountfromMetaData() const
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrproviderutils.cpp:2934]:
```suggestion
GIntBig QgsOgrLayer::GetTotalFeatureCountFromMetaData() const
```

File: src/core/providers/ogr/qgsogrproviderutils.cpp [hunk 5]
```diff
@@ -2909,6 +2927,41 @@ GIntBig QgsOgrLayer::GetApproxFeatureCount()
...
+{
+  // Don't quote column name (see https://trac.osgeo.org/gdal/ticket/5799#comment:9)
+  QByteArray layerName = OGR_L_GetName( hLayer );
+
+  QString driverName = GDALGetDriverShortName( GDALGetDatasetDriver( ds->hDS ) );
+  if ( driverName == QLatin1String( "GPKG" ) )
+  {
+    QByteArray sql( "SELECT feature_count FROM gpkg_ogr_contents WHERE " );
+    sql += "lower(table_name) = lower(" + QgsOgrProviderUtils::quotedIdentifier( layerName, driverName ) + ") LIMIT 2";
+    //sql += ")";
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrproviderutils.cpp:2944]:
```suggestion
```

**Review by rouault — APPROVED** [review]

**Pull request merged**